cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,63 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <random>
+
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/pcg_random.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+// Generates a random floating point number in range [0, 1).
+template <class Rng>
+double generate_uniform(Rng& rng) {
+  double range = static_cast<double>(Rng::max() - Rng::min()) + 1;
+  double value = static_cast<double>(rng() - Rng::min());
+  return value / range;
+}
+
+Status ExecRandom(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  static thread_local std::random_device rd;
+  static thread_local random::pcg64 gen(rd());
```

Comment:
`random_device` is implementation defined.
I think we should provide a kernel option to pass an initial seed and get reproduce-able output.

Reply:
The seed should probably be optional, however, so that people can have an automatic seeding from a "real" random source.

Reply:
I don't quite understand the state transition, let's say I provide seed s1, which seeds the generator. Does the generator keep using s1 until I provide another seed s2? How do I get back to using `random_device` or another source of entropy if I provided a seed once?

Looks like we'd either need a separate runtime function to set the seed and reset the generator or multiplex quite a bit of functionality in a single API function.

Reply:
Options are provided once for each call. There's no state transition.

One possible implementation is to provide two options. One option (bool) to select using random seed or preset seed, and another option (int) for the preset seed.
If the first option is true(random seed), we will use std::random_device to generate the initial seed, and ignore the second option.
If the first options is false(preset seed), then we will use the second option as the initial seed.

@pitrou please correct me if I'm wrong.

Looks current implementation generates only one scalar per call, so a seed may be a bit awkward to use. You have to choose a different seed for each call to get a different output.
The common use case is to generate a random array, it's desirable to provide a way to get same output for all calls on all platforms if we provide the same initial seed.

Reply:
Returning a scalar makes it weird, yes. If we make it return an array of double, we'd need a parameter to specify its size.

Reply:
You're right, we'll need an option to specify the size, which makes this function a bit unusual.

The options could look like this:
```c++
class RandomOptions : public FunctionOptions {
 public:
  static RandomOptions FromSystemRandom(int64_t length) {
    return RandomOptions{SystemRandom, length, 0};
  }
  static RandomOptions FromSeed(int64_t length, int64_t seed) {
    return RandomOptions{Seed, length, seed};
  }
  // Default constructor creates an invalid options struct, but is required for Python
  RandomOptions() {}

  enum Initializer { SystemRandom, Seed; };
  Initializer initializer = SystemRandom;
  int64_t length = -1;   // -1 means invalid
  int64_t seed = 0;
};
```

or perhaps:
```c++
class RandomOptions : public FunctionOptions {
 public:
  static RandomOptions FromSystemRandom(int64_t length) {
    return RandomOptions{SystemRandom(), length};
  }
  static RandomOptions FromSeed(int64_t length, int64_t seed) {
    return RandomOptions{Seed{seed}, length};
  }
  // Default constructor creates an invalid options struct, but is required for Python
  RandomOptions() {}

  struct Seed { int64_t seed; };
  struct SystemRandom {};
  using Initializer = util::variant<SystemRandom, Seed>;

  Initializer initializer = SystemRandom();
  int64_t length = -1;   // -1 means invalid
};
```

@bkietz Thoughts?

Reply:
It will a bit tricky how we plug these kinds of generation functions into an exec plan.  We will want  the size of the input batch to be mapped into the function options.  Everything we have today funnels through `arrow::compute::ExecuteScalarExpression` in `expression.cc`.  If we assume all nillary functions use some set of common options (we can tackle this when there is more than one nillary function) that might be enough.  We could just code something like...

```
  auto kernel = call->kernel;
  auto descrs = GetDescriptors(arguments);
  auto options = call->options.get();
  // -- new --
  if (arguments.size() == 0) {
    auto nillary_options = std::dynamic_pointer_cast<NillaryOptions>(options);
    nillary_options.size = input.length;
  }
```

Reply:
Maybe it shouldn't be a nullary function then? We could make the size an argument and leave the rest to the options.

Reply:
> Maybe it shouldn't be a nullary function then? We could make the size an argument and leave the rest to the options.

Passing size in as an argument would be more confusing I think.

Another option would be to make it a vector function that takes in the entire batch (possibly even a batch with zero columns) and then uses the size of the batch to create the output.  But, while that would make sense from a streaming execution standpoint, I think that would confuse any other direct consumer of the compute APIs.

My vote would still be size as an option.

Reply:
Right, so perhaps have a base `NullaryOptions` and derive from that in `RandomOptions`.